### PR TITLE
Implement label-based container discovery for Redis Dev Service

### LINKED ISSUE #18029: Implement shared Redis Server with DevServices
## Description

Ability to reuse a shared server launched by dev services across multiple Quarkus apps

## Implementation ideas

This should be label based discovery just like what we have for [Kafka](https://github.com/quarkusio/quarkus/pull/17837) and  [AMQP](https://github.com/quarkusio/quarkus/pull/17980)

### PR DESCRIPTION
**Pull request by patrox**
The PR started as a copy-paste solution for label-based container discovery for Redis dev service -
in order to handle: https://github.com/quarkusio/quarkus/issues/18029.

Based on these two PRs:

- Kafka: https://github.com/quarkusio/quarkus/pull/17837
- AMQP: https://github.com/quarkusio/quarkus/pull/17980

As many of you have noticed this required refactoring - so for that purpose I introduced two new classes:

- [ContainerLocator](https://github.com/quarkusio/quarkus/pull/18181/files#diff-46fcf1e20a4ba63f591741375828efe01cd293e303ff664663f4262476fb5238)
- [ContainerAddess](https://github.com/quarkusio/quarkus/pull/18181/files#diff-efd6f27826c11155e7d487d185a3eed8e0eaa230aa131ba02f87d1937da0ef06)

Both of them are located in a new module called simply: `Quarkus - DevServices - Common`.
The commits were not squashed (yet) for easier review.

### CODE REVIEW

**Comment by patrox** [comment]
@machi1990 @stuartwdouglas @cescoffier @Ladicek Can I please kindly ask you for a review?

**Review by stuartwdouglas — APPROVED** [review]

File: docs/src/main/asciidoc/redis.adoc
Comment (cescoffier) [inline docs/src/main/asciidoc/redis.adoc:92]: Can you extract the dev services for Redis documentation into its own file as we did for Kafka and AMQP.
We are refactoring the documentation to distinguish the "getting started" part from the "reference" part and the configuration of dev services is in the later part.

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java
Comment (cescoffier) [inline extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:3]: Unfortunately, most of the time it won't be enough.
We may need to add "extra" parameters. 

We can do that later.

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java
Comment (cescoffier) [inline extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:7]: can the port be `null`? If not, why `Integer` and not `int`?

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java
Comment (cescoffier) [inline extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:12]: What about just `getHost`.
First, it may not be an IP, but a name (like "localhost"), and from a user POV, we just need the host name (how is it resolved is irrelevant).

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java
Comment (cescoffier) [inline extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:21]: Should we add an optional scheme? Would that make sense?

**Review by machi1990 — COMMENTED** [review]
lgtm, thanks @patrox 

Can you squash the commits?

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java
Comment (cescoffier) [inline extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java:52]: why "locator"?

File: extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java
Comment (cescoffier) [inline extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:59]: Why these spaces?

File: extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java
Comment (cescoffier) [inline extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:61]: Why these spaces?

File: extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java
Comment (cescoffier) [inline extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:155]: Can this be extracted too?
Maybe not easily.

File: extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java
Comment (cescoffier) [inline extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:200]: What about having a container decorator that adds the label. 
At the moment we use a label-based mechanism, but this is likely going to change. 

(can be done in another PR).

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: docs/src/main/asciidoc/redis.adoc
Comment (patrox) [reply docs/src/main/asciidoc/redis.adoc:92]: Sure, I'll do that.

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java
Comment (patrox) [reply extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:3]: Yes, I'm aware of that. The current state of this class is to address the combined requirements for:

- Redis
- Kafka
- AMQP

Future cases might require additional fields.

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java
Comment (patrox) [reply extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:7]: Theoretically, it can, as the [ContainerPort#getPublicPort](https://github.com/docker-java/docker-java/blob/master/docker-java-api/src/main/java/com/github/dockerjava/api/model/ContainerPort.java#L67) is annoted with `CheckForNull`.

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java
Comment (patrox) [reply extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:12]: Yes, this is a better name.

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java
Comment (patrox) [reply extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:21]: At the moment it wasn't (isn't) needed - I wanted to keep it simple.

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java
Comment (patrox) [reply extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java:52]: This is how the class which encapsulates the logic is called (`Container Locator` - used to locate containers).

File: extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java
Comment (patrox) [reply extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:59]: I believe that this was autoformatted by my IntelliJ to align it with the other params.
So the `runTimeConfiguration` is aligned with `launchMode` above.

I can revert it as it's just formatting.

File: extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java
Comment (patrox) [reply extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:155]: I'll take a look at this as well, as it would further reduce the duplicated code.

**Comment by patrox** [comment]
> lgtm, thanks @patrox
> 
> Can you squash the commits?

Sure, they will be squashed.

**Comment by Ladicek** [comment]
Before I get to reviewing (which I'd really like to do per commit), what's the obsession with squashing commits? Clearly someone has made an effort to divide the work into smaller chunks that form a logical progression -- why would we want to lose that?

**Comment by machi1990** [comment]
> Before I get to reviewing (which I'd really like to do per commit), what's the obsession with squashing commits? Clearly someone has made an effort to divide the work into smaller chunks that form a logical progression -- why would we want to lose that?

It is not a obsession, but a suggestion. To me this boils down to two:
- Implementing the label based discovery on the Redis extension using the logic from existing extension
- Avoiding duplicates of the label-based discovery logic by extracting common code in a module and applying it through out. 

I am all +1 on having atomic commits, that can easily be cherry-picked and have the build pass.

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java
Comment (Ladicek) [reply extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java:52]: Except what you _found_ here is not a _container locator_; it's the _container_.

File: extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java
Comment (Ladicek) [reply extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:59]: `mvn clean compile` should fix that automatically (using the Quarkus formatter style).

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java
Comment (patrox) [reply extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java:52]: > Except what you _found_ here is not a _container locator_; it's the _container_.

Correct, the message will be changed to be more clear.

**Comment by Ladicek** [comment]
OK, I just went through the commits and the last one modifies multiple modules, whereas previously, each module was modified in a separate commit (which is very nice IMHO). In which case, squashing all the commits (except maybe the first one) would probably be best indeed. (Or percolate the changes from the last commit to each individual previous commit, though I can see how that's not the best use of one's time :-) )

**Comment by patrox** [comment]
> OK, I just went through the commits and the last one modifies multiple modules, whereas previously, each module was modified in a separate commit (which is very nice IMHO). In which case, squashing all the commits (except maybe the first one) would probably be best indeed. (Or percolate the changes from the last commit to each individual previous commit, though I can see how that's not the best use of one's time :-) )

About the squashing process - I usually like to create a PR with separate commits to make the review a bit easier.
But, before merging it, I prefer to squash them and leave just a small number of commits that are related to separate changes.

In this case, I would leave just two commits:

- Implement label-based container lookup for Redis Dev Service
- Introduce a common label-based container locator + refactoring

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java
Comment (patrox) [reply extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:7]: Correct, so I added an explicit check for this unlikely event, to be on the safe side.

File: extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java
Comment (patrox) [reply extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:61]: Fixed.

File: extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java
Comment (patrox) [reply extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:59]: Fixed.

File: extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java
Comment (patrox) [reply extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:155]: @cescoffier I see two possibilities how this can be extracted:

- in a smaller (simpler) scope we could extract the conditional logic for label-based container location (lines 155-160)
- in a bit broader scope we could extract the above __and__ also the other possibility, that is a creation of a new `Container` - which is done when the `shared` container is not used. (lines 155-172).

WDYT?

**Comment by patrox** [comment]
@cescoffier @Ladicek @machi1990 I addressed all of your comments, can I please ask you kindly for a 2nd round?

Please note that the commits will be squashed as discussed above - to 2 separate commits.

File: docs/src/main/asciidoc/redis.adoc
Comment (machi1990) [inline docs/src/main/asciidoc/redis.adoc:632]: Thanks for extracting this into a separate. 

This was previously linked in the beginning of the doc using the "dev-services" anchor. Do you mind updating the two `NOTE` to link to the added doc?

Comment (patrox) [reply docs/src/main/asciidoc/redis.adoc:632]: Sure, I'll do that.

Comment (patrox) [reply docs/src/main/asciidoc/redis.adoc:632]: @machi1990 Done.

**Review by machi1990 — APPROVED** [review]

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java
Comment (Ladicek) [inline extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:11]: Maybe that's just me, but the best signalization that `null` is forbidden would be to not accept `Integer`, but simply `int`.

File: docs/src/main/asciidoc/redis-dev-services.adoc
Comment (Ladicek) [inline docs/src/main/asciidoc/redis-dev-services.adoc:11]: Any reason to not call this "Dev Services for Redis"? That's how similar guides are called for Kafka, AMQP or Apicurio Registry.

Comment (patrox) [reply docs/src/main/asciidoc/redis-dev-services.adoc:11]: @Ladicek Actually this is a result of refactoring, as it's already called `Dev Service for Redis` - as this is how I named it (to be in line with i.e. Kafka guide).

The line which you are referring to is a left-over, so I'll remove it as it doesn't seem to be needed:

image

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java
Comment (patrox) [reply extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:11]: Yes, this is also an option - but this will require a check before passing the value to the ctor.
I can refactor it - it's not a problem.

File: extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java
Comment (cescoffier) [reply extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:155]: I would go with the first one.

**Review by cescoffier — CHANGES_REQUESTED** [review]
A minor comment still missing (some tiny refactoring), the rest looks great.

**Comment by patrox** [comment]
> A minor comment still missing (some tiny refactoring), the rest looks great.

@cescoffier I have this change already prepared locally, I'll push it a bit later.

File: extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java
Comment (patrox) [reply extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java]: @cescoffier I pushed the latest changes, generally, the `ContainerLocater` was made more "functional" by using `Optional` instead of bare `null`. This has shortened the code and additionally, it allowed to compose the method calls in a more fluent way.

File: docs/src/main/asciidoc/redis-dev-services.adoc
Comment (stuartwdouglas) [inline docs/src/main/asciidoc/redis-dev-services.adoc:11]: Shouldn't this just go in the main redis doc as a subsection?

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java
Comment (cescoffier) [inline extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java:36]: Minor: The 3 filters can be merged in a single one.

**Review by cescoffier — APPROVED** [review]

File: docs/src/main/asciidoc/redis-dev-services.adoc
Comment (cescoffier) [reply docs/src/main/asciidoc/redis-dev-services.adoc:11]: Kafka and AMQP already use this structure. The getting started contains admonitions referring to this file. Dev service configuration is more reference content than getting started content.

File: extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java
Comment (patrox) [reply extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java:36]: @cescoffier Yes, I did consider it, but I didn't want to make this line too long - as right now it's easier to read.
Actually now, when I saw this code again with a fresh head - I think it could be extracted to a predicate, so we will a single filter - which won't be too long at the same time.

**Review by gsmet — DISMISSED** [review]
Once the dust has settled, please try to squash the commits a bit to have proper semantic ones if possible.

**Comment by patrox** [comment]
> Once the dust has settled, please try to squash the commits a bit to have proper semantic ones if possible.

@gsmet  The commits were squashed and the branch was rebased to the latest ~~`master`~~, sorry `main`.

**Comment by patrox** [comment]
@gsmet @Ladicek I have squashed the commits so there are only two commits.
Can you please have a look and approve if there are no more comments from your side ?

**Review by Ladicek — APPROVED** [review]
I think this is OK, though there's certain school of thought (relatively prevalent on the Quarkus team) that prefers anonymous classes over lambdas because of memory footprint. Since the changes from anonymous classes to lambdas in this PR are in `deployment` modules and hence don't affect production builds, I think that should be fine :-)

**Comment by patrox** [comment]
> I think this is OK, though there's certain school of thought (relatively prevalent on the Quarkus team) that prefers anonymous classes over lambdas because of memory footprint. Since the changes from anonymous classes to lambdas in this PR are in `deployment` modules and hence don't affect production builds, I think that should be fine :-)

Thanks, @Ladicek for this insightful answer - I was actually wondering what's the rationale for such heavy usage of anonymous classes - and now I know.

**Comment by machi1990** [comment]
Thanks @patrox for the great work.

**Pull request merged**
